### PR TITLE
9548 saml soap backchannel logout

### PR DESCRIPTION
**Pull request by lagess**
Closes #9548

### CODE REVIEW

File: services/src/main/java/org/keycloak/protocol/saml/profile/util/Soap.java [hunk 1]
```diff
@@ -110,6 +111,19 @@ public static Document extractSoapMessage(SOAPMessage soapMessage) {
         }
     }
 
+    /**
+     * Get the first direct child that is an XML element
+     * (avoid returning text in case the first child is preceded by some new line)
+     */
+    public static Node getFirstChild(Node parent) {
```
Comment (basbeu) [inline services/src/main/java/org/keycloak/protocol/saml/profile/util/Soap.java:119]: I did not get the purpose of this method at the first read. Maybe the comment should more clear
And it seems to be not tested by a unit test

File: services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java [hunk 2]
```diff
@@ -748,6 +759,38 @@ protected Response buildLogoutResponse(UserSessionModel userSession, String logo
...
+                LogoutRequestType logoutRequest = createLogoutRequest(soapLogoutUrl, clientSession, client);
+                Document samlLogoutRequest = createBindingBuilder(samlClient, false).postBinding(SAML2Request.convert(logoutRequest)).getDocument();
+                SOAPMessage soapResponse = Soap.createMessage().addToBody(samlLogoutRequest).call(soapLogoutUrl);
+                Document logoutResponse = Soap.extractSoapMessage(soapResponse);
+                SAMLDocumentHolder samlDoc = SAML2Response.getSAML2ObjectFromDocument(logoutResponse);
+                if (!(samlDoc.getSamlObject() instanceof StatusResponseType)) {
+                    throw new RuntimeException("Logout response received from client is not an StatusResponseType");
+                }
+                StatusResponseType logoutStatus = (StatusResponseType)samlDoc.getSamlObject();
+                if (!logoutStatus.getStatus().getStatusCode().getValue().equals(JBossSAMLURIConstants.STATUS_SUCCESS.getUri())) {
```
Comment (basbeu) [inline services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java]: Is NPE impossible here ?

**Review by basbeu — APPROVED** [review]

**Review by fperot74 — APPROVED** [review]

File: services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java (hunk 2 above)
Comment (lagess) [reply services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java]: Could be indeed, but if there is an NPE, then the client has provided a response that is not valid according to the SAML protocol. I've slightly adapted the equality check to be a bit safer, and to do the same as an existing piece of code (https://github.com/keycloak/keycloak/blob/168734b817da2138548bacba973204e80f566b2c/services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java#L616)

File: services/src/main/java/org/keycloak/protocol/saml/profile/util/Soap.java (hunk 1 above)
Comment (lagess) [reply services/src/main/java/org/keycloak/protocol/saml/profile/util/Soap.java:119]: comment improved

**Pull request closed without merging**